boot/seal.go

```diff
@@ -93,8 +106,46 @@ func sealKeyToModeenv(key, saveKey secboot.EncryptionKey, model *asserts.Model,
 	return sealKeyToModeenvUsingSecboot(key, saveKey, model, modeenv)
 }
 
+func runKeySealRequests(key secboot.EncryptionKey) []secboot.SealKeyRequest {
+	return []secboot.SealKeyRequest{
+		{
+			Key:     key,
+			KeyFile: filepath.Join(InitramfsBootEncryptionKeyDir, "ubuntu-data.sealed-key"),
+		},
+	}
+}
+
+func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealKeyRequest {
+	return []secboot.SealKeyRequest{
+		{
+			Key:     key,
+			KeyFile: filepath.Join(InitramfsSeedEncryptionKeyDir, "ubuntu-data.recovery.sealed-key"),
+		},
+		{
+			Key:     saveKey,
+			KeyFile: filepath.Join(InitramfsSeedEncryptionKeyDir, "ubuntu-save.recovery.sealed-key"),
+		},
+	}
+}
+
 func sealKeyToModeenvUsingFdeSetupHook(key, saveKey secboot.EncryptionKey, model *asserts.Model, modeenv *Modeenv) error {
-	return fmt.Errorf("cannot use fde-setup hook yet")
+	// TODO: support full boot chains
+
+	for _, skr := range append(runKeySealRequests(key), fallbackKeySealRequests(key, saveKey)...) {
+		params := &FDESetupHookParams{
+			Key:     skr.Key,
+			KeyName: filepath.Base(skr.KeyFile),
```

Comment:
this doesn't look quite right,  we need to decide what exactly is a key name

Reply:
We can probably leave that as a TODO for now and not set it just yet

Reply:
This is no longer passed to the hook until we have a better idea what the keyname is. Alternatively we could maybe do something like:
```
diff --git a/boot/seal.go b/boot/seal.go
index a494fcd8ce..864397b693 100644
--- a/boot/seal.go
+++ b/boot/seal.go
@@ -110,6 +110,7 @@ func runKeySealRequests(key secboot.EncryptionKey) []secboot
.SealKeyRequest {
        return []secboot.SealKeyRequest{
                {
                        Key:     key,
+                       KeyName: "ubuntu-data",
                        KeyFile: filepath.Join(InitramfsBootEncryptionKeyDir, "ubuntu-data.sealed-key"),
                },
        }
@@ -119,10 +120,12 @@ func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealK
        return []secboot.SealKeyRequest{
                {
                        Key:     key,
+                       KeyName: "ubuntu-data",
                        KeyFile: filepath.Join(InitramfsSeedEncryptionKeyDir, "ubuntu-data.recovery.sealed-key"),
                },
                {
                        Key:     saveKey,
+                       KeyName: "ubuntu-save",
                        KeyFile: filepath.Join(InitramfsSeedEncryptionKeyDir, "ubuntu-save.recovery.sealed-key"),
                },
        }
@@ -133,10 +136,9 @@ func sealKeyToModeenvUsingFDESetupHook(key, saveKey secboot.EncryptionKey, model
 
        for _, skr := range append(runKeySealRequests(key), fallbackKeySealRequests(key, saveKey)...) {
                params := &FDESetupHookParams{
-                       Key: skr.Key,
-                       // TODO: decide what the right KeyName is
-                       // KeyName: filepath.Base(skr.KeyFile),
-                       Models: []*asserts.Model{model},
+                       Key:     skr.Key,
+                       KeyName: skr.KeyName,
+                       Models:  []*asserts.Model{model},
                }
                sealedKey, err := RunFDESetupHook("initial-setup", params)
                if err != nil {
diff --git a/secboot/secboot.go b/secboot/secboot.go
index 3c2149b126..cb85886719 100644
--- a/secboot/secboot.go
+++ b/secboot/secboot.go
@@ -56,6 +56,8 @@ func NewLoadChain(bf bootloader.BootFile, next ...*LoadChain) *LoadChain {
 type SealKeyRequest struct {
        // The key to seal
        Key EncryptionKey
+       // The keyName
+       KeyName string
        // The path to store the sealed key file
        KeyFile string
 }
```

Reply:
KeyName set like that would match fde-reveal-key usage, so would be ok.   KeyFile though gives the impression that the hook could read/write there which is not the case, also if/when we implement using Luks key slots we won't have a filename anymore.  We could have something like key-copy-name which is mostly for logging /debug purposes that we would set for now to the Base(key filename) perhaps?

Reply:
Thank you! I think this makes sense, I will clean this up and do it as a followup, this will be easier to review.